Fix Popup size on Android and iOS

cat0363 opened the pull request:
<!--
 Hello, and thank you for your interest in contributing to the .NET MAUI Toolkit! 

 Before you submit please check that this work relates to one of the following:
 - Bug fix
    If you haven't yet opened an Issue that reports the bug in detail, provides a reproduction sample, and has been verified + reproduced by a member of the .NET MAUI Toolkit core team, please do that before submitting a Pull Request.
 - Feature/Proposal
    If you haven't yet submitted a Proposal that has been Championed by a .NET MAUI core team member, please instead open a Discussion at https://github.com/communitytoolkit/maui/discussions/new where we can discuss the pros/cons of the feature and its implementation. 
 Any PR submitted that does not fit with the above options will be closed.
 -->

This PR corrects the Popup size to be calculated correctly when the Popup size is not explicitly specified.

 ### Description of Change ###

The process for getting the window size has been made into a method.

[src\CommunityToolkit.Maui.Core\Views\Popup\PopupExtensions.android.cs]

    static Size GetWindowSize(IWindowManager? windowManager, ViewGroup decorView)
    {
        ArgumentNullException.ThrowIfNull(windowManager);

        int windowWidth;
        int windowHeight;
        int statusBarHeight;
        int navigationBarHeight;

        if (OperatingSystem.IsAndroidVersionAtLeast((int)BuildVersionCodes.R))
        {
            var windowMetrics = windowManager.CurrentWindowMetrics;
            var windowInsets = windowMetrics.WindowInsets.GetInsetsIgnoringVisibility(WindowInsets.Type.SystemBars());
            windowWidth = windowMetrics.Bounds.Width();
            windowHeight = windowMetrics.Bounds.Height();
            statusBarHeight = windowInsets.Top;
            navigationBarHeight = windowHeight < windowWidth ? windowInsets.Left + windowInsets.Right : windowInsets.Bottom;
        }
        else if (windowManager.DefaultDisplay is null)
        {
            throw new InvalidOperationException($"{nameof(IWindowManager)}.{nameof(IWindowManager.DefaultDisplay)} cannot be null");
        }
        else
        {
            APoint realSize = new();
            APoint displaySize = new();
            ARect contentRect = new();

            windowManager.DefaultDisplay.GetRealSize(realSize);
            windowManager.DefaultDisplay.GetSize(displaySize);
            decorView.GetWindowVisibleDisplayFrame(contentRect);

            windowWidth = realSize.X;
            windowHeight = realSize.Y;
            statusBarHeight = contentRect.Top;

            navigationBarHeight = realSize.Y < realSize.X
                                                ? (realSize.X - displaySize.X)
                                                : (realSize.Y - displaySize.Y);
        }

        windowWidth = (windowWidth - (windowHeight < windowWidth ? navigationBarHeight : 0));
        windowHeight = (windowHeight - ((windowHeight < windowWidth ? 0 : navigationBarHeight) + statusBarHeight));

        return new Size(windowWidth, windowHeight);
    }

Then I modified the CalculateSizes method as follows:

[src\CommunityToolkit.Maui.Core\Views\Popup\PopupExtensions.android.cs]

    static void CalculateSizes(IPopup popup, Context context, Size windowSize, ref int realWidth, ref int realHeight, ref int realContentWidth, ref int realContentHeight)
    {
        ArgumentNullException.ThrowIfNull(popup.Content);

        var density = context.Resources?.DisplayMetrics?.Density ?? throw new InvalidOperationException($"Unable to determine density. {nameof(context.Resources.DisplayMetrics)} cannot be null");

        if (!popup.Size.IsZero)
        {
            realWidth = (int)context.ToPixels(popup.Size.Width);
            realHeight = (int)context.ToPixels(popup.Size.Height);
        }
        if (double.IsNaN(popup.Content.Width) || double.IsNaN(popup.Content.Height))
        {
            var size = popup.Content.Measure(windowSize.Width / density, windowSize.Height / density);
            realContentWidth = (int)context.ToPixels(size.Width);
            realContentHeight = (int)context.ToPixels(size.Height);
        }
        else
        {
            realContentWidth = (int)context.ToPixels(popup.Content.Width);
            realContentHeight = (int)context.ToPixels(popup.Content.Height);
        }

        realWidth = realWidth is 0 ? realContentWidth : realWidth;
        realHeight = realHeight is 0 ? realContentHeight : realHeight;

        if (realHeight is 0 || realWidth is 0)
        {
            realWidth = (int)(context.Resources?.DisplayMetrics?.WidthPixels * 0.8 ?? throw new InvalidOperationException($"Unable to determine width. {nameof(context.Resources.DisplayMetrics)} cannot be null"));
            realHeight = (int)(context.Resources?.DisplayMetrics?.HeightPixels * 0.6 ?? throw new InvalidOperationException($"Unable to determine height. {nameof(context.Resources.DisplayMetrics)} cannot be null"));
        }

        popup.Content.Measure(realWidth / density, realHeight / density);
    }

Changed to recalculate the size within the size of the Window when the Popup's Content size is not specified.
And, Changed to limit Popup size to Window size in SetSize method.

[src\CommunityToolkit.Maui.Core\Views\Popup\PopupExtensions.android.cs]

    realWidth = realWidth <= windowSize.Width ? realWidth : (int)windowSize.Width;
    realHeight = realHeight <= windowSize.Height ? realHeight : (int)windowSize.Height;
    window.SetLayout(realWidth, realHeight);

    var childLayoutParams = (FrameLayout.LayoutParams)(child.LayoutParameters ?? throw new InvalidOperationException($"{nameof(child.LayoutParameters)} cannot be null"));
    childLayoutParams.Width = realWidth;
    childLayoutParams.Height = realHeight;
    child.LayoutParameters = childLayoutParams;

    horizontalParams = realWidth;
    verticalParams = realHeight;

    var containerLayoutParams = new FrameLayout.LayoutParams(horizontalParams, verticalParams);

With these changes, the Popup's size is now correctly calculated when no explicit size is specified for the Popup.

 <!-- Describe your changes here. This only needs to be brief as the linked issues below will already cover the detailed changes. -->

 ### Linked Issues ###
 <!-- Provide links to issues here (#35 will link to issue number 35). Ensure that a GitHub issue was created for your bug/proposal and it has been approved/Championed. -->

 - Fixes #1353 
 - Fixes https://github.com/CommunityToolkit/Maui/issues/1416
- Fixes https://github.com/CommunityToolkit/Maui/issues/1427

 ### PR Checklist ###
 <!--
 Please check all the things you did here and double-check that you got it all, or state why you didn't do something.

 If anything is unclear please do ask :)
 -->
 - [x] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [ ] Has tests (if omitted, state reason in description)
 - [x] Has samples (if omitted, state reason in description)
 - [x] Rebased on top of `main` at time of PR
 - [x] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [ ] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pull/XYZ <!-- Replace XYZ with your docs PR #. The checkbox will be automatically checked once the docs PR has been merged -->


 ### Additional information ###

 <!-- 
 Please use this to aid the reviewer, this could include stating which platform(s) have been tested
 -->
 
Below are the verification results for Issue #1353.

<img width="250" src="https://github.com/CommunityToolkit/Maui/assets/125236133/23e7345f-f58b-47e2-b3f8-cc5d14ad0938" />

You can see that the label placed inside the Popup wraps as intended.

Even if you specify an explicit size for the Popups Size when the Popups content is only a Label, the label 
text was not wrapped before the fix, but after the fix, the label text is wrapped.

I layout the Popup as shown below and verified it.

    <toolkit:Popup xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
        xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
        xmlns:toolkit="http://schemas.microsoft.com/dotnet/2022/maui/toolkit"
        x:Class="MauiComm_IssuePopupWindows.PopupPage1" Size="100,100">

        <Label Text="1234567890ABCDEFGHIJKLMNOPQRSTUVWXYZ" TextColor="Black" />

    </toolkit:Popup>

Below are the verification results.

[When VerticalOptions is Start]
I changed the HorizontalOptions to:

<table>
  <tr>
    <td>[Start]</td>
    <td>[Center]</td>
    <td>[End]</td>
    <td>[Fill]</td>
  </tr>
  <tr>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/bbf9b5ae-9575-4b4f-9d53-f03bb8d8899e" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/ff5aea13-552d-42bd-9c55-3481cb4e7c8c" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/2b5eecb5-73e6-43c5-83c6-6afc95854f03" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/09b45886-fa98-43c7-831c-f0b5e3040890" /></td>
  </tr>
</table>

[When VerticalOptions is Center]
I changed the HorizontalOptions to:

<table>
  <tr>
    <td>[Start]</td>
    <td>[Center]</td>
    <td>[End]</td>
    <td>[Fill]</td>
  </tr>
  <tr>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/24351a51-a61f-49d3-b478-a4227cb270b9" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/e53d48ec-acf4-43ea-b167-4d2b5aaf70d6" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/60ff4f8b-52d8-405b-a948-26bb788c4ca1" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/298c7973-c006-45cd-b7a6-f4d75605f45e" /></td>
  </tr>
</table>

[When VerticalOptions is End]
I changed the HorizontalOptions to:

<table>
  <tr>
    <td>[Start]</td>
    <td>[Center]</td>
    <td>[End]</td>
    <td>[Fill]</td>
  </tr>
  <tr>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/41f1f5c9-7e96-42e9-9fa7-45c6f61a66fc" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/f9a818dc-c23c-4640-94f4-5ef3571a19bd" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/40e40c70-c87e-4b1f-9067-855df295b48b" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/f698e853-be75-428d-9919-d88850903afc" /></td>
  </tr>
</table>

[When VerticalOptions is Fill]
I changed the HorizontalOptions to:

<table>
  <tr>
    <td>[Start]</td>
    <td>[Center]</td>
    <td>[End]</td>
    <td>[Fill]</td>
  </tr>
  <tr>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/3ae88082-90d5-4a51-8e15-fb3bb0d62114" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/15a64e41-345e-4b37-9789-8bba7e6eacaf" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/174be100-63d2-43dc-88cb-e7cf06e3a6aa" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/4a6550df-42cc-4996-862e-2048f4402512" /></td>
  </tr>
</table>

Even if you explicitly specify the size of the Popup, you can see that the label inside the Popup is wrapped.

cat0363 commented:
I need to make a similar fix on the iOS side, so can I include the fix on the iOS side in this PR?

pictos commented:
When not specifying the size, does Popup respects the LayoutOptions (Center, Fill, Start, End)?

cat0363 commented:
> When not specifying the size, does Popup respects the LayoutOptions (Center, Fill, Start, End)?

@pictos , I'll fix it again because the PR I created doesn't respect the LayoutOptions when the Popup size is unspecified.
Thanks for the good question.

cat0363 commented:
@pictos , Since iOS has a similar issue, is it okay to include the fix in this PR?
If there is no problem with including it, I will change the title later.
For Windows, I will include the fix in PR #1350. This is because the modifications conflict.

cat0363 commented:
I made the following changes so that the LayoutOptions are respected when the Popup size is unspecified.

The below are changes on the Android side.

[src\CommunityToolkit.Maui.Core\Views\Popup\PopupExtensions.android.cs]

    static void CalculateSizes(IPopup popup, Context context, Size windowSize, ref int realWidth, ref int realHeight, ref int realContentWidth, ref int realContentHeight)
    {
        ArgumentNullException.ThrowIfNull(popup.Content);

        var density = context.Resources?.DisplayMetrics?.Density ?? throw new InvalidOperationException($"Unable to determine density. {nameof(context.Resources.DisplayMetrics)} cannot be null");

        if (!popup.Size.IsZero)
        {
            realWidth = (int)context.ToPixels(popup.Size.Width);
            realHeight = (int)context.ToPixels(popup.Size.Height);
        }
        else
        {
            if (double.IsNaN(popup.Content.Width) || double.IsNaN(popup.Content.Height))
            {
                var size = popup.Content.Measure(windowSize.Width / density, windowSize.Height / density);
                realContentWidth = (int)context.ToPixels(size.Width);
                realContentHeight = (int)context.ToPixels(size.Height);

                if (double.IsNaN(popup.Content.Width))
                {
                    realContentWidth = popup.HorizontalOptions == LayoutAlignment.Fill ? (int)windowSize.Width : realContentWidth;
                }
                if (double.IsNaN(popup.Content.Height))
                {
                    realContentHeight = popup.VerticalOptions == LayoutAlignment.Fill ? (int)windowSize.Height : realContentHeight;
                }
            }
            else
            {
                realContentWidth = (int)context.ToPixels(popup.Content.Width);
                realContentHeight = (int)context.ToPixels(popup.Content.Height);
            }
        }

        realWidth = Math.Min(realWidth is 0 ? realContentWidth : realWidth, (int)windowSize.Width);
        realHeight = Math.Min(realHeight is 0 ? realContentHeight : realHeight, (int)windowSize.Height);

        if (realHeight is 0 || realWidth is 0)
        {
            realWidth = (int)(context.Resources?.DisplayMetrics?.WidthPixels * 0.8 ?? throw new InvalidOperationException($"Unable to determine width. {nameof(context.Resources.DisplayMetrics)} cannot be null"));
            realHeight = (int)(context.Resources?.DisplayMetrics?.HeightPixels * 0.6 ?? throw new InvalidOperationException($"Unable to determine height. {nameof(context.Resources.DisplayMetrics)} cannot be null"));
        }

        popup.Content.Measure(realWidth / density, realHeight / density);
    }

Since I moved the code from the SetSize method to the CalculateSizes method, I removed the two lines 
below from the SetSize method.

    realWidth = realWidth <= windowSize.Width ? realWidth : (int)windowSize.Width;
    realHeight = realHeight <= windowSize.Height ? realHeight : (int)windowSize.Height;

The below are changes on the iOS side.

[src\CommunityToolkit.Maui.Core\Views\Popup\PopupExtensions.macios.cs]

    public static void SetSize(this MauiPopup mauiPopup, in IPopup popup)
    {
        ArgumentNullException.ThrowIfNull(popup.Content);

        CGRect frame;

        if (mauiPopup.ViewController?.View?.Window is UIWindow window)
        {
            frame = window.Frame;
        }
        else
        {
            frame = UIScreen.MainScreen.Bounds;
        }

        CGSize currentSize;

        if (!popup.Size.IsZero)
        {
            currentSize = new CGSize(popup.Size.Width, popup.Size.Height);
        }
        else
        {
            if (double.IsNaN(popup.Content.Width) || double.IsNaN(popup.Content.Height))
            {
                var content = popup.Content.ToPlatform(popup.Handler?.MauiContext ?? throw new InvalidOperationException($"{nameof(popup.Handler.MauiContext)} Cannot Be Null"));
                var contentSize = content.SizeThatFits(new CGSize(frame.Width, frame.Height));
                var width = contentSize.Width;
                var height = contentSize.Height;

                if (double.IsNaN(popup.Content.Width))
                {
                    width = popup.HorizontalOptions == Microsoft.Maui.Primitives.LayoutAlignment.Fill ? frame.Size.Width : width;
                }
                if (double.IsNaN(popup.Content.Height))
                {
                    height = popup.VerticalOptions == Microsoft.Maui.Primitives.LayoutAlignment.Fill ? frame.Size.Height : height;
                }

                currentSize = new CGSize(width, height);
            }
            else
            {
                currentSize = new CGSize(popup.Content.Width, popup.Content.Height);
            }
        }

        currentSize.Width = NMath.Min(currentSize.Width, frame.Size.Width);
        currentSize.Height = NMath.Min(currentSize.Height, frame.Size.Height);
        mauiPopup.PreferredContentSize = currentSize;
    }

Changed code to respect LayoutOptions when popup size is unspecified and popup content size is 
unspecified on both Android and iOS.

Verified whether LayoutOptions are respected when Popup size is not specified.
The layout of the Popup used for verification is as follows.

    <toolkit:Popup xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
        xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
        xmlns:toolkit="http://schemas.microsoft.com/dotnet/2022/maui/toolkit"
        x:Class="MauiComm_IssuePopupWindows.PopupPage1">

        <Label Text="1234567890ABCDEFGHIJKLMNOPQRSTUVWXYZ" TextColor="Black" HorizontalOptions="Start" />

    </toolkit:Popup>

Below are the verification results on Android.

[When VerticalOptions is Start]
I changed the HorizontalOptions to:

<table>
  <tr>
    <td>[Start]</td>
    <td>[Center]</td>
    <td>[End]</td>
    <td>[Fill]</td>
  </tr>
  <tr>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/7333e385-77a8-42c7-801e-de583a4c3e1d" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/91a23fe6-e84f-450a-81cd-a77523fe6bb3" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/35b5e259-d806-40da-9fd4-a450cc669b60" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/46787ea7-8b47-4b8d-a9b3-4cdae69fdf08" /></td>
  </tr>
</table>

[When VerticalOptions is Center]
I changed the HorizontalOptions to:

<table>
  <tr>
    <td>[Start]</td>
    <td>[Center]</td>
    <td>[End]</td>
    <td>[Fill]</td>
  </tr>
  <tr>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/84b9c6fd-1fa1-4112-9eb8-bd08db3e6044" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/1fec7213-ffda-435d-9b17-bbafff79123f" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/89e60848-a247-4131-81fa-e6626ce5c8e4" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/7260eb3d-5b45-41d4-a1c4-ed0038f3f515" /></td>
  </tr>
</table>

[When VerticalOptions is End]
I changed the HorizontalOptions to:

<table>
  <tr>
    <td>[Start]</td>
    <td>[Center]</td>
    <td>[End]</td>
    <td>[Fill]</td>
  </tr>
  <tr>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/93f576e2-1b11-4998-9a8f-681241608e7c" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/ea32219e-47c6-4ffe-b48f-a6a3476a4b40" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/cefdfdb8-806b-492d-b0bc-22da2cbeecce" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/c8f976fa-6465-4919-a311-cd6f3dedeebb" /></td>
  </tr>
</table>

[When VerticalOptions is Fill]
I changed the HorizontalOptions to:

<table>
  <tr>
    <td>[Start]</td>
    <td>[Center]</td>
    <td>[End]</td>
    <td>[Fill]</td>
  </tr>
  <tr>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/0a9d04b6-4362-4232-861e-d76b306ab546" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/ddf84c85-f5ef-485e-bff4-c982b2232aad" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/d0a47a4e-e587-4ae5-ab03-0ca0ef180b56" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/2e9ad94c-ba8e-4752-9687-d772152ff3ec" /></td>
  </tr>
</table>

Below are the verification results on iOS.

[When VerticalOptions is Start]
I changed the HorizontalOptions to:

<table>
  <tr>
    <td>[Start]</td>
    <td>[Center]</td>
    <td>[End]</td>
    <td>[Fill]</td>
  </tr>
  <tr>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/fb874c00-b415-4352-8bdb-f266a475c121" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/63a79037-5746-48f6-8b3d-5df7c62a5f36" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/705d12ef-8047-4624-b703-eb6ef4fb6e70" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/63806914-8fe8-4d61-8e68-b411b24a91ea" /></td>
  </tr>
</table>

[When VerticalOptions is Center]
I changed the HorizontalOptions to:

<table>
  <tr>
    <td>[Start]</td>
    <td>[Center]</td>
    <td>[End]</td>
    <td>[Fill]</td>
  </tr>
  <tr>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/de7e64db-ee43-430b-ba52-dc6d28ae4d37" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/b71002ca-5018-40ce-ada7-f0f80d88e8e5" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/b2d3e452-92e2-49d6-8f27-bf903cf29445" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/8da0c6ff-daa1-4418-9c0e-79ef5d48c6ca" /></td>
  </tr>
</table>

[When VerticalOptions is End]
I changed the HorizontalOptions to:

<table>
  <tr>
    <td>[Start]</td>
    <td>[Center]</td>
    <td>[End]</td>
    <td>[Fill]</td>
  </tr>
  <tr>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/1ccaafa8-0174-4179-9b90-1a5f7093a77b" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/c48a331a-cdb5-41f1-bbe7-0d13285cb6af" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/bba70747-a2e2-4172-bea8-86d39f6ed6ba" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/a861553d-1a98-4eaa-9001-0f51d5864be4" /></td>
  </tr>
</table>

[When VerticalOptions is Fill]
I changed the HorizontalOptions to:

<table>
  <tr>
    <td>[Start]</td>
    <td>[Center]</td>
    <td>[End]</td>
    <td>[Fill]</td>
  </tr>
  <tr>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/a6542755-d4cf-4591-9531-236305f677bf" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/eb9b9eb3-34ca-455e-b32c-f04f9b97a5ad" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/69cefae6-06af-4966-88ed-7c540d38bbca" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/ef623a64-907c-408b-8594-dc973beb2106" /></td>
  </tr>
</table>

For both Android and iOS, you can see that LayoutOptions are respected when the Popup size 
is unspecified and the Popup Content size is unspecified.

And, I laid out the Popup like this:

    <toolkit:Popup xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
        xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
        xmlns:toolkit="http://schemas.microsoft.com/dotnet/2022/maui/toolkit"
        x:Class="MauiComm_VerifyPopupStyle.Popup6" VerticalOptions="Center" HorizontalOptions="Start">

        <VerticalStackLayout>
            <Label Text="Popup" FontSize="20"/>
            <BoxView HeightRequest="3" Color="AliceBlue" />        
            <Label x:Name="txt1" Text="Short String" />
            <Label x:Name="txt2" Text="Long String1 Long String2 Long String3 Long String4 Long String5 Long String6 Long String7 Long String8 Long String9 Long String10 Long String" />
        </VerticalStackLayout>

    </toolkit:Popup>

Below is the execution result on iOS.

<img width="250" src="https://github.com/CommunityToolkit/Maui/assets/125236133/e64229ee-0e0f-40af-96e0-01c7cea42603" />

You can see that the labels are wrapped like in Android.

cat0363 commented:
Below are additional validation results.

    <toolkit:Popup xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
        xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
        xmlns:toolkit="http://schemas.microsoft.com/dotnet/2022/maui/toolkit"
        x:Class="MauiComm_IssuePopupWindows.PopupPage2">

        <Grid WidthRequest="100">
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="Start" HorizontalOptions="Start" BackgroundColor="Blue" />
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="Start" HorizontalOptions="End" BackgroundColor="Blue" />
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="End" HorizontalOptions="Start" BackgroundColor="Blue" />
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="End" HorizontalOptions="End" BackgroundColor="Blue" />
        </Grid>

    </toolkit:Popup>

Below are the verification results on Android.

[When VerticalOptions is Start]
I changed the HorizontalOptions to:

<table>
  <tr>
    <td>[Start]</td>
    <td>[Center]</td>
    <td>[End]</td>
    <td>[Fill]</td>
  </tr>
  <tr>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/dd549e2b-7903-4f2e-bfdf-453834aaca68" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/a1f01a7f-ad86-41b2-a643-84afe57123b6" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/c45aea46-006c-4c57-9472-63f92423c2cd" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/de2e5ac7-7c48-41b2-b9b2-47afa42fe0b3" /></td>
  </tr>
</table>

[When VerticalOptions is Center]
I changed the HorizontalOptions to:

<table>
  <tr>
    <td>[Start]</td>
    <td>[Center]</td>
    <td>[End]</td>
    <td>[Fill]</td>
  </tr>
  <tr>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/352c6546-96a7-4c07-b0b2-3c68bd427d1a" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/23d0f2ba-2070-4403-9afc-3a0f5bc835af" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/333c7493-b87f-402b-90d4-29680ced9c07" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/83e6d213-4ae5-485c-9540-0590e5064f6c" /></td>
  </tr>
</table>

[When VerticalOptions is End]
I changed the HorizontalOptions to:

<table>
  <tr>
    <td>[Start]</td>
    <td>[Center]</td>
    <td>[End]</td>
    <td>[Fill]</td>
  </tr>
  <tr>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/0ce8881a-2618-4ccc-9849-e19e39d8688c" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/3a435d33-e69c-49c9-8109-00bbbe8f693b" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/f4f38e15-b325-46e0-a493-ef9d29155072" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/16267253-bf4c-4364-9e58-987aed809ca7" /></td>
  </tr>
</table>

[When VerticalOptions is Fill]
I changed the HorizontalOptions to:

<table>
  <tr>
    <td>[Start]</td>
    <td>[Center]</td>
    <td>[End]</td>
    <td>[Fill]</td>
  </tr>
  <tr>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/58f94f34-e883-4472-8e3b-97f188e8d342" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/7ede4989-65a3-46bb-be38-662bfce30eaf" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/a6279aa6-200d-478e-b2b7-4164634751a1" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/64ea0e1b-bd1b-4e9a-871d-363a59c51ca3" /></td>
  </tr>
</table>

Below are the verification results on iOS.

[When VerticalOptions is Start]
I changed the HorizontalOptions to:

<table>
  <tr>
    <td>[Start]</td>
    <td>[Center]</td>
    <td>[End]</td>
    <td>[Fill]</td>
  </tr>
  <tr>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/91fed703-8822-470c-bde9-70e25c10df95" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/08847956-bb78-49df-b069-5e6e5127679a" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/dabe2bff-341c-40f7-8e16-971251c67fc0" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/2a11f956-6c8e-487c-8e08-f99b6682842c" /></td>
  </tr>
</table>

[When VerticalOptions is Center]
I changed the HorizontalOptions to:

<table>
  <tr>
    <td>[Start]</td>
    <td>[Center]</td>
    <td>[End]</td>
    <td>[Fill]</td>
  </tr>
  <tr>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/06a5ea1b-896e-446f-b15a-330d8a6121d0" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/4586bbfd-ad5d-4c3b-8c74-db8671340474" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/4c1fd40d-7698-479f-b77e-af711ca0c761" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/5d38322b-84b1-4517-a078-a44aefc27bab" /></td>
  </tr>
</table>

[When VerticalOptions is End]
I changed the HorizontalOptions to:

<table>
  <tr>
    <td>[Start]</td>
    <td>[Center]</td>
    <td>[End]</td>
    <td>[Fill]</td>
  </tr>
  <tr>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/9248a072-2674-4df7-a4ac-d9f75dc5db26" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/ac8485a1-2236-4863-b1b0-03175d32deab" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/8707bbcd-d944-4b79-80c8-3f7d2bc1f6df" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/baa0f2e7-6a51-40f6-91c0-7842e6f3cec4" /></td>
  </tr>
</table>

[When VerticalOptions is Fill]
I changed the HorizontalOptions to:

<table>
  <tr>
    <td>[Start]</td>
    <td>[Center]</td>
    <td>[End]</td>
    <td>[Fill]</td>
  </tr>
  <tr>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/854bf6cc-e955-4a95-acbd-9399a8d64523" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/f9e17af3-daf7-41a4-9149-31fbb8d4fa75" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/5560935f-cebb-4f61-8120-87f29ee76bd8" /></td>
    <td><img width="200" src="https://github.com/CommunityToolkit/Maui/assets/125236133/47879287-aa21-45ce-9f24-927554716a19" /></td>
  </tr>
</table>

You can see that the VerticalOptions are respected because the height is unspecified.
I won't post the test results, but it will now respect HorizontalOptions when width is unspecified.

cat0363 commented:
@pictos , Windows test results are posted below.
https://github.com/CommunityToolkit/Maui/pull/1350#discussion_r1302506429

cat0363 commented:
Following the comments below, I have made similar fixes to this PR as well.https://github.com/CommunityToolkit/Maui/pull/1350#discussion_r1303444004

pictos commented on src/CommunityToolkit.Maui.Core/Views/Popup/PopupExtensions.android.cs:
```diff
@@ -227,35 +165,98 @@ public static void SetSize(this Dialog dialog, in IPopup popup, in AView contain
+			}
+			else
+			{
+				APoint realSize = new();
+				APoint displaySize = new();
+				ARect contentRect = new();
```
Can we verify if those values will change? I'm quite sure that `displaySize` will not change during the application lifetime, so we can cache it and avoid calling the Android APIs every time, this will improve perf. a lot. If they still change, and we need to do the calls to `widowManager` and `decorView`, can we reuse the objects, instead of creating new ones every time?

cat0363 replied:
@pictos , Thank you for your review.
The variables realSize, displaySize and contentRect change with screen rotation.
Fixed not to create an object for each call by storing it in a static variable.

cat0363 commented:
@pictos , @VladislavAntonyuk 
I want to fix the position when popups are overlaid on iOS.
If the first Popup's VerticalOptions and HorizontalOptions are not Start, the second Popup will be displayed 
in an unintended position. Once the fix is ​​complete, I'll post it here.

cat0363 commented:
I'm sorry for the change after approval. Below are the corrections.

[src\CommunityToolkit.Maui.Core\Handlers\Popup\PopupHandler.macios.cs]

    public static async void MapOnClosed(PopupHandler handler, IPopup view, object? result)
    {
        var pc = handler.PlatformView.PresentationController;
        if (pc is not null)
        {
            if (pc.PresentedViewController is UIViewController pvc)
            {
                await pvc.DismissViewControllerAsync(true);
            }
        }

        view.HandlerCompleteTCS.TrySetResult();

        handler.DisconnectHandler(handler.PlatformView);
    }

This fix is ​​taken from PR #1369.
The above is a modification to close the Popup in order from the last displayed Popup when displaying Popups overlaid.

[src\CommunityToolkit.Maui.Core\Views\Popup\PopupExtensions.macios.cs]

    public static void SetLayout(this MauiPopup mauiPopup, in IPopup popup)
    {
        if (mauiPopup.View is null)
        {
            return;
        }

        CGRect frame;

        if (mauiPopup.ViewController?.View?.Window is UIWindow window)
        {
            frame = window.Frame;
        }
        else
        {
            frame = UIScreen.MainScreen.Bounds;
        }

        if (mauiPopup.PopoverPresentationController is null)
        {
            throw new InvalidOperationException("PopoverPresentationController cannot be null");
        }

        if (popup.Anchor is null)
        {
            omit ...

            if (mauiPopup.ViewController?.PopoverPresentationController is UIPopoverPresentationController popoverPresentationController)
            {
                mauiPopup.PopoverPresentationController.SourceView = popoverPresentationController.SourceView;
            }
            mauiPopup.PopoverPresentationController.SourceRect = new CGRect(originX, originY, 0, 0);
            mauiPopup.PopoverPresentationController.PermittedArrowDirections = 0;
            // From the point of view of usability, the top, bottom, left, and right values of UIEdgeInsets cannot all be 0.
            // If you specify 0 for the top, bottom, left, and right of UIEdgeInsets, the default margins will be added, so 
            // specify a value as close to 0 here as possible.
            mauiPopup.PopoverPresentationController.PopoverLayoutMargins = new UIEdgeInsets(0.0001f, 0.0001f, 0.0001f, 0.0001f);
        }
        else
        {
            var view = popup.Anchor.ToPlatform(popup.Handler?.MauiContext ?? throw new InvalidOperationException($"{nameof(popup.Handler.MauiContext)} Cannot Be Null"));
            mauiPopup.PopoverPresentationController.SourceView = view;
            mauiPopup.PopoverPresentationController.SourceRect = view.Bounds;
        }
    }

The above is a correction of the base position of the second and subsequent Popups.
The Popup's SourceView is set to the View that is the original display source of the Popup.

The verification results before and after modification are shown below.

The 1st Popup has HorizontalOptions="End" and VerticalOptions="End" set.
The 2nd Popup has HorizontalOptions="Center" and VerticalOptions="Center" set.
The 3rd Popup has HorizontalOptions="Start" and VerticalOptions="Start" set.

<table>
  <tr>
    <td>[Before Fix]</td>
    <td>[After Fix]</td>
  </tr>
  <tr>
    <td><video controls width="250" src="https://github.com/CommunityToolkit/Maui/assets/125236133/58ee353e-4f45-4f5c-99cd-6822b0da84fa" /></td>
    <td><video controls width="250" src="https://github.com/CommunityToolkit/Maui/assets/125236133/ece05bec-dc59-43da-ab5e-6d806a0cd031" /></td>
  </tr>
</table>

Before the modification, when the third Popup was closed, the second Popup was also closed, but after the modification, only the third Popup was closed.

Before the correction, the next Popup was displayed starting from the position of the previous Popup, but after the correction, the Popup is displayed starting from the View that is the display source of the Popup.

With the overlay fix in PR #1369, the second and subsequent Popups will no longer be transparent.

cat0363 commented:
If PR #1409 is applied, this PR will no longer work as intended ...
It's a good fix for MacCatalyst, but it seems to have a big impact on iOS.
Popups can no longer be displayed nested. The view displaying the Popup disappears at the same time the Popup is displayed.

pictos commented:
@cat0363 if we make that change just for mac, does it work? Or is there a way to make both codes work?

cat0363 commented:
@pictos , Thank you for your reply.
If you apply PR #1409 for Mac only, the problem I described will not occur on iOS.
However, it is unclear whether it will work as intended on Mac if this PR is additionally 
applied with PR #1409 applied.

I think PR #1409 is a nice fix that makes sense, but it reduces flexibility on iOS.
I haven't found a way to make both codes work at the moment.

pictos commented:
@cat0363, so you're saying that the quick fix is to apply that fix just on macOS, right? If so we can open an issue and a PR for that (if you want to do it go ahead otherwise I can try to find a time this week to perform the change)

cat0363 commented:
@pictos , If PR #1409 is applied, it seems that this PR does not work as intended, or rather that the previous operation is not guaranteed. 

>@cat0363, so you're saying that the quick fix is to apply that fix just on macOS, right? If so we can open an issue and a PR for that (if you want to do it go ahead otherwise I can try to find a time this week to perform the change)

I'm not hoping for a quick fix, but this is how it currently works, is it okay?
At least on iOS this will be a problem.

https://github.com/CommunityToolkit/Maui/assets/125236133/82ae7e8b-e366-4df0-b7ed-191cdc6aeff3

The above verification video has nothing to do with the verification of this PR.

When PR #1409 is applied, the following warning appears when displaying Popup.

    [Presentation] Presenting view controller <CommunityToolkit_Maui_Core_Views_MauiPopup: 0x7f97cff7b010> from detached view controller <Microsoft_Maui_Platform_PageViewController: 0x7f97cffa0590> is discouraged.

The above warning is issued in the following places.

[src\CommunityToolkit.Maui.Core\Views\Popup\MauiPopup.macios.cs]

    void AddToCurrentPageViewController(UIViewController viewController)
    {
        viewController.PresentViewController(this, true, null);
    }

cat0363 commented:
@pictos , I found a way to make both work by changing PR #1409, so I'll post it as a bug.

babenkovitaliy commented:
> @pictos , I found a way to make both work by changing PR #1409, so I'll post it as a bug.

I periodically randomly check my submitted PRs and found this thread referencing my last PR. If my PR is affecting iOS, what if we go with something along these lines:

```
	/// <summary>
	/// Method to initialize the native implementation.
	/// </summary>
	/// <param name="element">An instance of <see cref="IPopup"/>.</param>
	[MemberNotNull(nameof(VirtualView), nameof(ViewController))]
	public void SetElement(IPopup element)
	{
		if (element.Parent?.Handler is not PageHandler)
		{
			throw new InvalidOperationException($"The {nameof(element.Parent)} must be of type {typeof(PageHandler)}.");
		}

		VirtualView = element;
		ModalPresentationStyle = UIModalPresentationStyle.Popover;

		_ = View ?? throw new InvalidOperationException($"{nameof(View)} cannot be null.");
		_ = VirtualView ?? throw new InvalidOperationException($"{nameof(VirtualView)} cannot be null.");

#if MACCATALYST
		var pagehandler = VirtualView.Parent.Handler as PageHandler;
		var rootViewController = pagehandler?.ViewController ?? WindowStateManager.Default.GetCurrentUIViewController() ?? throw new InvalidOperationException($"{nameof(PageHandler.ViewController)} cannot be null.");
#else
		var rootViewController = WindowStateManager.Default.GetCurrentUIViewController() ?? throw new InvalidOperationException($"{nameof(PageHandler.ViewController)} cannot be null.");
#endif

		ViewController ??= rootViewController;
		SetDimmingBackgroundEffect();
	}
```

By adding `#if MACCATALYST`, Popups in MacOS will get handled one way and then Popups in iOS will get handled how it was previously (inside the `#else` preprocessor directive).

cat0363 commented:
@babenkovitaliy , Thank you for posting this matter.
Let's discuss [here](https://github.com/CommunityToolkit/Maui/issues/1416).

brminnick commented:
Hey Gang! Thanks for your hard work! 

Just checking in - do we need to resolve https://github.com/CommunityToolkit/Maui/issues/1416 in this PR, or will we be opening a separate PR for it?

cat0363 commented:
@brminnick , Currently, the following warning is output and the iOS side does not function as before, 
so I would like to target the fix for PR #1409 only to MacCatalyst. 

    [Presentation] Presenting view controller <CommunityToolkit_Maui_Core_Views_MauiPopup: 0x7f97cff7b010> from detached view controller <Microsoft_Maui_Platform_PageViewController: 0x7f97cffa0590> is discouraged.

I would like to include the fix for Issue #1416 in this PR. Below are the changes.

[src\CommunityToolkit.Maui.Core\Views\Popup\MauiPopup.macios.cs]

    public void SetElement(IPopup element)
    {
        if (element.Parent?.Handler is not PageHandler)
        {
            throw new InvalidOperationException($"The {nameof(element.Parent)} must be of type {typeof(PageHandler)}.");
        }

        VirtualView = element;
        ModalPresentationStyle = UIModalPresentationStyle.Popover;

        _ = View ?? throw new InvalidOperationException($"{nameof(View)} cannot be null.");
        _ = VirtualView ?? throw new InvalidOperationException($"{nameof(VirtualView)} cannot be null.");

    #if MACCATALYST
        var pagehandler = VirtualView.Parent.Handler as PageHandler;
        var rootViewController = pagehandler?.ViewController ?? WindowStateManager.Default.GetCurrentUIViewController() ?? throw new InvalidOperationException($"{nameof(PageHandler.ViewController)} cannot be null.");
    #else
        var rootViewController = WindowStateManager.Default.GetCurrentUIViewController() ?? throw new InvalidOperationException($"{nameof(PageHandler.ViewController)} cannot be null.");
    #endif

        ViewController ??= rootViewController;
        SetDimmingBackgroundEffect();
    }

If I can find a better solution, I would like to create a new PR then. Thank you.

brminnick reviewed (changes requested):
Thanks @cat0363! FYI - I added the sample you created in your reproduction to the official sample app. Great work! 



This is really really close! I confirmed that everything looks good on iOS + Android.

But on macOS, I'm seeing strange behavior in the `CustomSizeAndPositionPopupPage`:
- When `VerticalOptions == End`, the Popup doesn't appear.
- When `HorizontalOptions == End`, the Popup appears outside of the macOS window
![ScreenFlow](https://github.com/CommunityToolkit/Maui/assets/13558917/82b4d10d-40f8-47ab-8ad8-d8a7b87308df)

VladislavAntonyuk commented:
Seems like the issue in calculating the offset. Popup is moved half of its size to the right

cat0363 commented:
@brminnick , Thank you for verifying with MacCatalyst.
When I verified the operation of PR #1409, I also noticed that the Popup position on MacCatalyst was incorrect.
By taking into account the height of the title bar and the height of the navigation bar, the vertical position will be 
close to what you expect. Regarding the horizontal position, it works as expected considering the offset of 20, 
but I don't see a clear reason for this 20. If you specify Start for HorizontalOptions, the position of the Popup will 
strangely shift so that it does not cover the close button, so this may be due to that shift.

systemMinimumLayoutMargins may be related, but I'm not sure ...

cat0363 commented:
I will reconsider including the rounding process when the Popup size exceeds the screen size.

cat0363 commented:
It seems that MacCatalyst does not allow Popup to completely cover the caller's screen as it does on iOS.
I'm not sure if that's the MacCatalyst spec. Even if the horizontal position is 0, some margin is always provided.
This is why magic numbers exist.

cat0363 commented:
In the case of MacCatalyst, there is a margin that is always taken on the left side, so modify it to take that 
margin around the Popup. It probably looks prettier that way.

cat0363 commented:
For MacCatalyst, I modified it so that a margin of 18 is taken around the Popup by default.
This 18 is the minimum margin on the left side of the Popup in MacCatalyst.

    #if MACCATALYST
        static nfloat popupMargin = 18f;
    #endif

You may need to look into the MacCatalyst specifications to find out why a minimum margin of 18 is required.
I haven't investigated that far.

cat0363 commented:
I have uploaded the fix for Issue #1423.
I fixed the size calculation when the Popup size is specified.

    static void CalculateSizes(IPopup popup, Context context, Size windowSize, ref int realWidth, ref int realHeight, ref int realContentWidth, ref int realContentHeight)
    {
        ArgumentNullException.ThrowIfNull(popup.Content);

        var density = context.Resources?.DisplayMetrics?.Density ?? throw new InvalidOperationException($"Unable to determine density. {nameof(context.Resources.DisplayMetrics)} cannot be null");

        if (popup.Size.IsZero)
        {
            if (double.IsNaN(popup.Content.Width) || double.IsNaN(popup.Content.Height))
            {
                var size = popup.Content.Measure(windowSize.Width / density, windowSize.Height / density);
                realContentWidth = (int)context.ToPixels(size.Width);
                realContentHeight = (int)context.ToPixels(size.Height);

                if (double.IsNaN(popup.Content.Width))
                {
                    realContentWidth = popup.HorizontalOptions == LayoutAlignment.Fill ? (int)windowSize.Width : realContentWidth;
                }
                if (double.IsNaN(popup.Content.Height))
                {
                    realContentHeight = popup.VerticalOptions == LayoutAlignment.Fill ? (int)windowSize.Height : realContentHeight;
                }
            }
            else
            {
                realContentWidth = (int)context.ToPixels(popup.Content.Width);
                realContentHeight = (int)context.ToPixels(popup.Content.Height);
            }
        }
        else
        {
            realWidth = (int)context.ToPixels(popup.Size.Width);
            realHeight = (int)context.ToPixels(popup.Size.Height);

            var size = popup.Content.Measure(popup.Size.Width, popup.Size.Height);
            realContentWidth = (int)context.ToPixels(size.Width);
            realContentHeight = (int)context.ToPixels(size.Height);							
        }
    
        realWidth = Math.Min(realWidth is 0 ? realContentWidth : realWidth, (int)windowSize.Width);
        realHeight = Math.Min(realHeight is 0 ? realContentHeight : realHeight, (int)windowSize.Height);

        if (realHeight is 0 || realWidth is 0)
        {
            realWidth = (int)(context.Resources?.DisplayMetrics?.WidthPixels * 0.8 ?? throw new InvalidOperationException($"Unable to determine width. {nameof(context.Resources.DisplayMetrics)} cannot be null"));
            realHeight = (int)(context.Resources?.DisplayMetrics?.HeightPixels * 0.6 ?? throw new InvalidOperationException($"Unable to determine height. {nameof(context.Resources.DisplayMetrics)} cannot be null"));
        }
    }

Below is a video of the verification results.

https://github.com/CommunityToolkit/Maui/assets/125236133/abb45454-3a9b-4ee5-aa07-5e56af74af6a

Added squares to the four corners of Popup in the reproduction code of Issue #1423.

I tried rotating the screen and verifying it.

<table>
  <tr>
    <td>[Before Rotation]</td>
    <td>[After Rotation]</td>
  </tr>
  <tr>
    <td><img width="250" src="https://github.com/CommunityToolkit/Maui/assets/125236133/21b13a6b-df5f-4845-9288-761650774113" /></td>
    <td><img width="250" src="https://github.com/CommunityToolkit/Maui/assets/125236133/38822cff-e169-4849-8c15-612dd401cd8e" /></td>
  </tr>
</table>

cat0363 commented:
Below is a verification video using MacCatalyst.

[Width = 100, Height = 100]

https://github.com/CommunityToolkit/Maui/assets/125236133/0858a706-6765-42fe-99dd-052426b57e18

[Width = 1000, Height = 100]

https://github.com/CommunityToolkit/Maui/assets/125236133/45684609-d0ee-435c-aebd-88a2449d67d5

[Width = 100, Height = 1000]

https://github.com/CommunityToolkit/Maui/assets/125236133/60593730-d38f-43d3-af1b-8b9919470ff0

[Width = 1000, Height = 1000]

https://github.com/CommunityToolkit/Maui/assets/125236133/d72d9a33-e73f-43b6-be26-179616eeacff

brminnick reviewed (approved):
Awesome!! Thank you so much Kenji for your always amazing work 💯
